src/olmo_core/data/source_mixture.py

```diff
@@ -0,0 +1,303 @@
+        if self.target_ratio:
+            if not 0 <= self.target_ratio <= 1:
+                raise OLMoConfigurationError("target_ratio must be in the range [0, 1]")
+            if not 0 <= self.max_source_fraction <= 1:
+                raise OLMoConfigurationError("max_source_fraction must be in the range [0, 1]")
+            if self.max_source_fraction < self.target_ratio:
+                raise OLMoConfigurationError("max_source_fraction must be >= target_ratio")
+
+        if not self.paths:
+            raise OLMoConfigurationError("paths must not be empty")
+
+        if not 0 <= self.max_source_fraction <= 1:
+            raise OLMoConfigurationError("max_source_fraction must be in the range [0, 1]")
+
+
+@dataclass
+class SourceTokenDetails:
+    """
+    A class to hold intermediate selection details for a mixture source.
+    """
+
+    config: SourceMixtureConfig
+    population: int
+    num_selected: int
+
+    def for_table(self, max_tokens: int) -> Dict:
+        return {
+            "source_name": self.config.source_name,
+            "source_population": f"{self.population:.2e}",
+            "num_selected": f"{self.num_selected:.2e}",
+            "target_ratio": str(self.config.target_ratio),
+            "max_repetion_ratio": str(self.config.max_repetition_ratio),
+            "max_source_fraction": str(self.config.max_source_fraction),
+            "observed_source_ratio": f"{(self.num_selected / self.population):.4}",
+            "observed_global_ratio": f"{(self.num_selected / max_tokens):.4}",
+        }
+
+
+@dataclass
+class SourcePathTokens:
+    path: PathOrStr
+    tokens: int
+
+
+@dataclass
+class SourceMixtureOutcome:
+    name: str
+    path_tokens: List[SourcePathTokens]
+
+
+@dataclass
+class SourceMixtureDataset:
+    """
+    A dataset consisting of a fractionalized mixture of data sources.
+    """
+
+    seed: int
+    sources: List[SourceMixtureOutcome]
+
+    def to_index(self) -> Dict[Tuple[str, int], int]:
+        """
+        Convert the dataset to an indexed array of dict((int, path), int).
+        """
+        return {
+            (str(outcome.path), idx): outcome.tokens
+            for idx, outcome in enumerate(
+                list(chain.from_iterable([outcome.path_tokens for outcome in self.sources]))
+            )
+        }
+
+    def to_paths(self) -> List[PathOrStr]:
+        """
+        Convert the dataset to a list of paths while maintaining stable ordering.
+        """
+        return [
+            item.path
+            for item in list(chain.from_iterable([outcome.path_tokens for outcome in self.sources]))
+        ]
+
+
+@dataclass
+class SourceMixtureDatasetConfig(Config):
+    """
+    A configuration class for building a dataset from a fractionalized mixture of sources.
+    """
+
+    max_tokens: int
+    source_configs: List[SourceMixtureConfig]
+    sequence_length: int
+    dtype: NumpyDatasetDType
+    processes: int = 1
+    seed: int = 42
```

Comment:
Could you add doc strings for these attributes? Even if most are obvious, our documentation build process won't show attributes that aren't documented.